builder/dockerfile/parser/parser.go

```diff
@@ -107,29 +108,31 @@ func (d *Directive) setEscapeToken(s string) error {
 	return nil
 }
 
-// processLine looks for a parser directive '# escapeToken=<char>. Parser
-// directives must precede any builder instruction or other comments, and cannot
-// be repeated.
-func (d *Directive) processLine(line string) error {
+// possibleParserDirective looks for one or more parser directives '# escapeToken=<char>' and
+// '# platform=<string>'. Parser directives must precede any builder instruction
+// or other comments, and cannot be repeated.
+func (d *Directive) possibleParserDirective(line string) error {
 	if d.processingComplete {
 		return nil
 	}
-	// Processing is finished after the first call
-	defer func() { d.processingComplete = true }()
 
+	logrus.Debugf("[BUILDER] considering parser directive in %q", line)
 	tecMatch := tokenEscapeCommand.FindStringSubmatch(strings.ToLower(line))
-	if len(tecMatch) == 0 {
-		return nil
-	}
-	if d.escapeSeen == true {
-		return errors.New("only one escape parser directive can be used")
-	}
-	for i, n := range tokenEscapeCommand.SubexpNames() {
-		if n == "escapechar" {
-			d.escapeSeen = true
-			return d.setEscapeToken(tecMatch[i])
+	if len(tecMatch) != 0 {
+		if d.escapeSeen == true {
+			return errors.New("only one escape parser directive can be used")
+		}
+		for i, n := range tokenEscapeCommand.SubexpNames() {
+			if n == "escapechar" {
+				logrus.Debugf("[BUILDER] escape set to '%s'", tecMatch[i])
+				d.escapeSeen = true
+				return d.setEscapeToken(tecMatch[i])
+			}
 		}
 	}
+
+	logrus.Debugf("[BUILDER] no more parser directives allowed from this point")
+	d.processingComplete = true
```

Comment:
The `defer func()` was in the removed side of the diff. I was suggesting that maybe removing it might be the only change required here, but I think I understand better now. I was assuming that all directives would match the same regex.

> I was thinking that adding a second directive would be something

Do you think it would make sense to have all directives share the same form (parsable by the same regex)? That would let us do something roughly like this:

```go

directiveKey, directiveValue := parseDirectiveLine(strings.ToLower(line))
if d.seenDirectives[directiveKey] {
    return errors.Errorf("only one %s parser directive can be used", directiveKey)
}

switch directiveKey {
case "escape":
    ...
case "otherdirective":
    ...
default:
    logrus.Debugf("[BUILDER] no more parser directives allowed from this point")
    d.processingComplete = true
    return nil
}

d.seenDirectives[directiveKey] = true
return nil
```